tools/shared/src/main/scala/org/scalajs/core/tools/linker/analyzer/Analyzer.scala

```diff
@@ -182,22 +191,29 @@ private final class Analyzer(semantics: Semantics,
 
     var nonExistent: Boolean = false
 
-    /** Ensures that this class and its dependencies are linked.
-     *
-     *  @throws CyclicDependencyException if this class is already linking
-     */
-    def linkClasses(): Unit = {
-      if (_linking)
-        throw CyclicDependencyException(encodedName :: Nil)
+    /** Ensures that this class and its dependencies are linked. */
+    def linkClasses(): Unit =
+      linkClasses(linkingFrom = null)
 
+    private def linkClasses(linkingFrom: ClassInfo): Unit = {
       if (!_linked) {
-        _linking = true
-        try {
+        if (_linkingFrom == null)
+          _linkingFrom = linkingFrom
+
+        if (!_linking) {
+          _linking = true
           linkClassesImpl()
-        } catch {
-          case CyclicDependencyException(chain) =>
-            throw CyclicDependencyException(encodedName :: chain)
+        } else {
+          def makeCycle(from: ClassInfo, acc: List[ClassInfo]): List[ClassInfo] = {
+            if (from == linkingFrom) from :: acc
+            else makeCycle(from._linkingFrom, from :: acc)
+          }
```

Comment:
I don't think this does the right thing. It walks up the dependency tree that has already been established and then checks for the `linkingFrom`. There is no guarantee we reach that:

Consier: `A <- B <- C <- B`. When we reach `B` for the second time, we get the following invocation order:

```
makeCycle(B, Nil)
makeCycle(A, B :: Nil)
makeCycle(null, A :: B :: Nil)
NPE
```

The proper invocation is `makeCycle(linkingFrom, Nil)` and the abort condition is `from eq this`:

```
makeCycle(C, Nil)
makeCycle(B, C :: Nil)
--> B :: C :: Nil
```

This is probably also the reason why you need to set `_linkingFrom` even if `_linking`. You can then "detect" loops if the loop involves the root linking node.